ws/api/endpoint_test.go

```diff
@@ -0,0 +1,189 @@
+	if subtopic != "" {
+		subtopicPart = fmt.Sprintf("/%s", subtopic)
+	}
+	if header {
+		return fmt.Sprintf("%s/channels/%s/messages%s", u, chanID, subtopicPart), nil
+	}
+
+	return fmt.Sprintf("%s/channels/%s/messages%s?authorization=%s", u, chanID, subtopicPart, thingKey), nil
+}
+
+func handshake(tsURL, chanID, subtopic, thingKey string, addHeader bool) (*websocket.Conn, *http.Response, error) {
+	header := http.Header{}
+	if addHeader {
+		header.Add("Authorization", thingKey)
+	}
+
+	url, _ := makeURL(tsURL, chanID, subtopic, thingKey, addHeader)
+	conn, res, errRet := websocket.DefaultDialer.Dial(url, header)
+
+	return conn, res, errRet
+}
+
+func TestHandshake(t *testing.T) {
+	thingsClient := httpmock.NewThingsClient(map[string]string{thingKey: chanID})
+	svc := newService(thingsClient)
+	ts := newHTTPServer(svc)
+	defer ts.Close()
+
+	cases := []struct {
+		name     string
+		chanID   string
+		subtopic string
+		header   bool
+		thingKey string
+		status   int
+		err      error
+		msg      []byte
+	}{
+		{
+			name:     "connect and send message",
+			chanID:   id,
+			subtopic: "",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect to empty channel",
+			chanID:   "",
+			subtopic: "",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusBadRequest,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect with empty thingKey",
+			chanID:   id,
+			subtopic: "",
+			header:   true,
+			thingKey: "",
+			status:   http.StatusForbidden,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect and send message with thingKey as query parameter",
+			chanID:   id,
+			subtopic: "",
+			header:   false,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message that cannot be published",
+			chanID:   id,
+			subtopic: "",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect and send message to subtopic",
+			chanID:   id,
+			subtopic: "subtopic",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to subtopic with invalid name",
+			chanID:   id,
+			subtopic: "sub/a*b/topic",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusBadRequest,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to nested subtopic",
+			chanID:   id,
+			subtopic: "subtopic/nested",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to all subtopics",
+			chanID:   id,
+			subtopic: ">",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+	}
+
+	for _, tt := range cases {
```

Comment:
What does `tt` stand for? Use `tc` (test case) for the sake of consistency with other tests.